Add ReorderableLists to `MessageFieldView` attachments (#1425)

IluhinDotPro opened the pull request:
Resolves #1425 

## Synopsis

`Attachments` attached cannot be sorted. Also their order is not the order that `ChatMessage` will display those attachments in.




## Solution

- Split `attachments` by to list (image/video and other files). 
- Added ReorderableList to both attachments lists

- Modifed `_addAttachment` in `MessageFieldController` for nore correctly add new `Attachment` to `attachments` (images/video first, files than)




## Checklist

- Created PR:
    - [x] In [draft mode][l:1]
    - [x] Name contains issue reference
    - [x] Has type and `k::` labels applied
- Before [review][l:4]:
    - [x] Documentation is updated (if required)
    - [x] Tests are updated (if required)
    - [x] Changes conform [code style][l:2]
    - [x] [CHANGELOG entry][l:3] is added (if required)
    - [x] FCM (final commit message) is posted or updated
    - [x] [Draft mode][l:1] is removed
- [ ] [Review][l:4] is completed and changes are approved
    - [ ] FCM (final commit message) is approved
- Before merge:
    - [ ] Milestone is set
    - [ ] PR's name and description are correct and up-to-date
    - [ ] All temporary labels are removed




[l:1]: https://help.github.com/en/articles/about-pull-requests#draft-pull-requests
[l:2]: https://github.com/team113/messenger/blob/main/CONTRIBUTING.md#code-style
[l:3]: https://github.com/team113/messenger/blob/main/CHANGELOG.md
[l:4]: https://help.github.com/en/articles/reviewing-changes-in-pull-requests

IluhinDotPro committed:
Add ReorderableLists to `MessageFieldView` attachments (#1425)

SleepySquash committed:
Merge remote-tracking branch 'origin/main' into 1425-implement-attachments-reordering-in-messagefieldview-by-long-tap

SleepySquash commented on lib/ui/page/home/page/chat/message_field/controller.dart:
```diff
@@ -572,7 +572,28 @@ class MessageFieldController extends GetxController {
         var attachment = LocalAttachment(file, status: SendingStatus.sending);
-        attachments.add(MapEntry(GlobalKey(), attachment));
+
+        // If attachment is video or image insert it to end of img/video list and before documents, overwise simply add to end of attachments
```
https://dart.dev/effective-dart/style#prefer-lines-80-characters-or-fewer

80 characters per line is a global standard limit: https://softwareengineering.stackexchange.com/questions/148677/why-is-80-characters-the-standard-limit-for-code-width

Plus don't forget, please, to end sentences with dots: https://dart.dev/effective-dart/documentation#do-format-comments-like-sentences

SleepySquash commented on lib/ui/page/home/page/chat/message_field/view.dart:
```diff
@@ -1020,6 +1116,26 @@ class MessageFieldView extends StatelessWidget {
     );
   }
+
+  /// Check that [entry] attachment is image or video
```
`Check` isn't a verb in third-person form. Effective Dart suggests to documents methods the other way: https://dart.dev/effective-dart/documentation#prefer-starting-comments-of-a-function-or-method-with-third-person-verbs-if-its-main-purpose-is-a-side-effect

SleepySquash commented on lib/ui/page/home/page/chat/message_field/view.dart:
```diff
@@ -422,12 +430,101 @@ class MessageFieldView extends StatelessWidget {
+                                      width:
+                                          (attachmentSize + 4) *
+                                          imgVideosAttachments.length,
+                                      child: ReorderableList(
```
Can't we add, please, some kind of UI/UX feedback to the end user when reordering begins? I suggest to perhaps add a small subtle shadow, if possible, when item gets into "reorder mode" (long tap gesture is captured by `ReorderableDelayedDragStartListener`). And a vibration feedback  when reorder starts or happens? We have a helper method to do vibration feedback in `PlatformUtils` variable - please, take a look!

SleepySquash reviewed (changes requested):
I've left some notes. Take a look, please!

IluhinDotPro committed:
Fix PR (#1517) issues (#1425)

- format code and comments with common style
- rename `imgVideosAttachments` to `mediaAttachments`
- add `AttachmentExtension` on `Attachment`
- change `attachmentSize` to static const `_attachmentSize`
- add UI/UX feedback when reorder was start and stop

IluhinDotPro committed:
Remove unused `attachmentsScrollController`

SleepySquash commented on lib/ui/page/home/page/chat/message_field/controller.dart:
```diff
@@ -209,6 +209,8 @@ class MessageFieldController extends GetxController {
   /// registered.
   bool _handlersRegistered = false;
 
+  final ScrollController attachmentsScrollController = ScrollController();
```
https://github.com/team113/messenger/blob/main/CONTRIBUTING.md#documentation-1
https://github.com/team113/messenger/blob/main/CONTRIBUTING.md#classes-constructors-fields-and-methods-ordering

SleepySquash commented on lib/api/backend/extension/attachment.dart:
```diff
@@ -0,0 +1,37 @@
+
+/// Extension adding media or file checks on [Attachment].
+extension AttachmentExtension on Attachment {
+  /// Indicates whether the [Attachment] is image
```
Why did you decide to add dots only to some sentences? Does this sentence differ somehow from others, where you placed the dots?

SleepySquash reviewed (changes requested):
You shouldn't hurry, please. Before sending PR for a review, do a self-check. It's better to take an hour or more just to be extra sure that you're doing what you intended to do.

IluhinDotPro committed:
Move `AttachmentExtension` to `lib\domain\extension\attachment.dart` (#1425)

- format code and comments with common style

SleepySquash reviewed (changes requested):
Few questions - take a look, please

IluhinDotPro committed:
Merge branch 'main' into 1425-implement-attachments-reordering-in-messagefieldview-by-long-tap

IluhinDotPro committed:
Merge branch 'main' of https://github.com/team113/messenger into 1425-implement-attachments-reordering-in-messagefieldview-by-long-tap

SleepySquash reviewed (changes requested):
Excuse me, have you seen the previous comments I've left in a review last week? https://github.com/team113/messenger/pull/1517#pullrequestreview-3510048422

All the comments are marked as "resolved", and this PR has been sent for a the review, however it seems that you neither did anything told in the comments nor wrote any arguments in those discussions as to why you don't agree with all of this.

Am I missing something? Or why did you resolve all the comments without doing or telling anything?

IluhinDotPro committed:
Format code and comments with common style (#1517, #1425)
